### ISSUE TITLE
Ensure Partition Specs can only contain primitive types

### ISSUE THREAD

**Issue by c-thiel** [apache/iceberg-rust#718]
Following the discussion here: https://github.com/apache/iceberg-rust/pull/349#discussion_r1855682480 and the [spec](https://iceberg.apache.org/spec/#partitioning), partitions must only produce primitive types.

This is currently not guranteed, neither by deserialization nor for `add_unbound_field`. At least adding a `void` transform can result in complex types.

**Comment by jagdeeps91** [comment]
I have a pull request for this issue: https://github.com/apache/iceberg-rust/pull/1285
This just removes this if/else check: https://github.com/apache/iceberg-rust/blob/main/crates/iceberg/src/spec/partition.rs#L582

As per spec the source column should be primitive type for all Transforms. 

@c-thiel As original author, please let me know if this check should not be removed.

This is my first PR to iceberg-rust, would be very happy to get feedback!

**Comment by jagdeeps91** [comment]
I have published a pull request, which removes an if check. Now the code validates that the source column type is Primitive for all transforms including Void:
https://github.com/apache/iceberg-rust/pull/1285

@c-thiel since you authored the original code, let me know if this if check should be kept:
https://github.com/apache/iceberg-rust/blob/main/crates/iceberg/src/spec/partition.rs#L582

This is my first pull request to iceberg-rust, looking for feedback!

**Comment by liurenjie1024** [comment]
There is a reason for skipping check for void transform: https://github.com/apache/iceberg/blob/3469cf30f2d839763813596e3f0c66ed16189c9a/api/src/main/java/org/apache/iceberg/PartitionSpec.java#L632-L636

**Comment by reneleonhardt** [comment]
@liurenjie1024 Could you summarize what the vision is so someone can implement it?
Christian said only produce primitive types, you said at least void Transform should remain.
And then "check each transform supports targeting type".
So what's it gonna be? 😄

**Comment by liurenjie1024** [comment]
> [@liurenjie1024](https://github.com/liurenjie1024) Could you summarize what the vision is so someone can implement it? Christian said only produce primitive types, you said at least void Transform should remain. And then "check each transform supports targeting type". So what's it gonna be? 😄

I think we should close this issue as not planned as the comment in java clearly clarified why we need to add this workaround. I'll close this for now, and feel free to reopen it when necessary.
